btree/btree.c

```diff
@@ -5814,68 +6238,318 @@ static bool vkvv_isfit(struct slot *slot)
+	int                  incoming_vsize    =
+				m0_vec_count(&slot->s_rec.r_val.ov_vec);
+	int                  i;
 
-	void             *start_dir_addr = vkvv_get_dir_addr(slot->s_node);
-	void             *key_addr       = vkvv_key(slot->s_node, count);
-	void             *val_addr       = vkvv_val(slot->s_node, count);
-	void             *end_dir_addr   = start_dir_addr +
-					   sizeof(struct dir_rec) * (count + 2);
+	if (!IS_EMBEDDED_INDIRECT(slot->s_node))
+		return false;
 
-	if (vkvv_crctype_get(slot->s_node) == M0_BCT_BTREE_ENC_RAW_HASH)
-		vsize += CRC_VALUE_SIZE;
+	if (vkvv_level(slot->s_node) == 0 &&
+	    vkvv_crctype_get(slot->s_node) == M0_BCT_BTREE_ENC_RAW_HASH)
+		incoming_vsize += CRC_VALUE_SIZE;
 
-	if (key_addr + ksize > start_dir_addr ||
-	    val_addr - vsize < end_dir_addr)
+	for (i = 0; i < h->vkvv_dir_entries; i++) {
+		total_avail_ksize += (dir[i].alloc_key_size - dir[i].key_size);
+		total_avail_vsize += (dir[i].alloc_val_size - dir[i].val_size);
+	}
+
+	if ((total_avail_ksize + total_avail_vsize) >= (incoming_ksize +
+	     incoming_vsize + sizeof(*dir)))
 		return true;
 	else
 		return false;
 }
 
-/**
- * @brief This function will move directory if the current space for values or
- *        keys overlaps with the space occupied by directory.
- */
-static void vkvv_move_dir(struct slot *slot)
+static void sort_dir(struct vkvv_dir_rec *dir, int count)
 {
-	struct vkvv_head *h              = vkvv_data(slot->s_node);
-	int               count          = h->vkvv_used;
-	uint32_t          ksize          =
-			m0_vec_count(&slot->s_rec.r_key.k_data.ov_vec);
-	uint32_t          vsize          =
+	int i, j;
+	struct vkvv_dir_rec item;
+	for (i = 1; i < count; i++) {
+		item = dir[i];
+		j = i - 1;
+		while (j >= 0 && dir[j].key_offset > item.key_offset) {
+			dir[j+1] = dir[j];
+			j = j -1;
+		}
+		dir[j+1] = item;
+	}
+}
+
+static bool is_rec_with_hole(struct vkvv_dir_rec entry)
+{
+	if(entry.key_size != 0 && entry.val_size != 0 &&
+	   ((entry.alloc_key_size - entry.key_size > 0) ||
+	    (entry.alloc_val_size - entry.val_size > 0)))
+		return true;
+	else
+		return false;
+}
+
+static bool is_rec_without_hole(struct vkvv_dir_rec entry)
+{
+	if(entry.key_size != 0 && entry.val_size != 0 &&
+	   entry.alloc_key_size - entry.key_size == 0 &&
+	   entry.alloc_val_size - entry.val_size == 0)
+		return true;
+	else
+		return false;
+}
+
+static bool is_rec_free_frag(struct vkvv_dir_rec entry)
+{
+	if(entry.key_size == 0 && entry.val_size == 0) {
+		M0_ASSERT(entry.alloc_key_size >= 0 ||
+			  entry.alloc_val_size >= 0);
+		return true;
+	}
+	else
+		return false;
+}
+
+static int get_idx_of_rec(const struct nd *node, struct vkvv_dir_rec entry)
+{
+	struct vkvv_head    *h   = vkvv_data(node);
+	struct vkvv_dir_rec *dir = vkvv_dir_get(node);
+	int                  i;
+
+	for(i = 0; i < h->vkvv_dir_entries; i++) {
+		if (dir[i].key_offset == entry.key_offset &&
+		    dir[i].val_offset == entry.val_offset)
+			break;
```

Comment:
No ned to check both the offsets. Instead, we can check only one offset and Assert the second offset. It would be more understandable.
if (dir[i].key_offset == entry.key_offset )
{
                 M0_ASSERT(dir[i].val_offset == entry.val_offset);
	         break;
}